polygeist/tools/cgeist/Lib/CGCall.cc

```diff
@@ -1517,17 +1518,48 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
   }
 
   /// If the callee is part of the SYCL namespace, we may not want the
-  /// GetOrCreateMLIRFunction to add this FuncOp to the functionsToEmit dequeu,
+  /// GetOrCreateMLIRFunction to add this FuncOp to the functionsToEmit deque,
   /// since we will create it's equivalent with SYCL operations. Please note
   /// that we still generate some functions that we need for lowering some
   /// sycl op.  Therefore, in those case, we set ShouldEmit back to "true" by
   /// looking them up in our "registry" of supported functions.
-  auto ShouldEmit =
-      !mlirclang::isNamespaceSYCL(callee->getEnclosingNamespaceContext());
-  std::string name;
-  MLIRScanner::getMangledFuncName(name, callee, Glob.getCGM());
-  if (isSupportedFunctions(name))
+  bool isSyclFunc =
+      mlirclang::isNamespaceSYCL(callee->getEnclosingNamespaceContext());
+  bool ShouldEmit = !isSyclFunc;
+
+  std::string mangledName;
+  MLIRScanner::getMangledFuncName(mangledName, callee, Glob.getCGM());
+
+  if (isSupportedFunctions(mangledName))
     ShouldEmit = true;
+
+  // If the caller is a SYCL kernel or device function, then the callee should
+  // also be a device function.
+  auto findCaller = [&](clang::CallExpr *expr) {
+    clang::DynTypedNodeList nodeList =
+        Glob.getCGM().getContext().getParents(*expr);
+    while (!nodeList.empty()) {
+      clang::DynTypedNode parentNode = nodeList[0];
+
+      if (const FunctionDecl *parent = parentNode.get<FunctionDecl>())
+        return parent;
+
+      nodeList = Glob.getCGM().getContext().getParents(parentNode);
+    }
+
+    llvm_unreachable("cannot find caller");
+    return (const FunctionDecl *)nullptr;
+  };
+  const FunctionDecl *caller = findCaller(expr);
+  bool isCallerDevice = caller->hasAttr<clang::SYCLKernelAttr>() ||
+                        caller->hasAttr<clang::SYCLDeviceAttr>();
+  if (isCallerDevice) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Adding device attribute to " << mangledName << "\n");
+    const_cast<FunctionDecl *>(callee)->addAttr(
+        SYCLDeviceAttr::CreateImplicit(Glob.getCGM().getContext()));
+  }
+
```

Comment:
I don't think we need this after #6916 and #6920 (also, I found those a better solution).

Reply:
Agree, I will rebase.